// now/build/lib/get-files.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.docker = exports.npm = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

/**
 * Returns a list of files in the given
 * directory that are subject to be
 * synchronized for npm.
 *
 * @param {String} full path to directory
 * @param {String} contents of `package.json` to avoid lookup
 * @param {Object} options:
 *  - `limit` {Number|null} byte limit
 *  - `debug` {Boolean} warn upon ignore
 * @return {Array} comprehensive list of paths to sync
 */

var npm = exports.npm = function () {
  var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(path, pkg) {
    var _ref2 = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

    var _ref2$limit = _ref2.limit;
    var limit = _ref2$limit === undefined ? null : _ref2$limit;
    var _ref2$debug = _ref2.debug;
    var debug = _ref2$debug === undefined ? false : _ref2$debug;
    var whitelist, search_, search, npmIgnore, filter, prefixLength, accepts, files;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            whitelist = pkg.now && pkg.now.files ? pkg.now.files : pkg.files;

            // the package.json `files` whitelist still
            // honors ignores: https://docs.npmjs.com/files/package.json#files

            search_ = whitelist || ['.'];

            // always include the "main" file

            if (pkg.main) {
              search_.push(pkg.main);
            }

            // convert all filenames into absolute paths
            search = search_.map(function (file) {
              return asAbsolute(file, path);
            });

            // compile list of ignored patterns and files

            _context.next = 6;
            return maybeRead((0, _path.resolve)(path, '.npmignore'), null);

          case 6:
            npmIgnore = _context.sent;
            _context.t0 = (0, _ignore2.default)();
            _context.t1 = _ignored2.default + '\n';

            if (!(null != npmIgnore)) {
              _context.next = 13;
              break;
            }

            _context.t2 = npmIgnore;
            _context.next = 16;
            break;

          case 13:
            _context.next = 15;
            return maybeRead((0, _path.resolve)(path, '.gitignore'));

          case 15:
            _context.t2 = _context.sent;

          case 16:
            _context.t3 = _context.t2;
            _context.t4 = clearRelative(_context.t3);
            _context.t5 = _context.t1 + _context.t4;
            filter = _context.t0.add.call(_context.t0, _context.t5).createFilter();
            prefixLength = path.length + 1;

            accepts = function accepts(file) {
              var relativePath = file.substr(prefixLength);
              if ('' === relativePath) return true;
              var accepted = filter(relativePath);
              if (!accepted && debug) {
                console.log('> [debug] ignoring "%s"', file);
              }
              return accepted;
            };

            // locate files


            if (debug) console.time('> [debug] locating files ' + path);
            _context.next = 25;
            return explode(search, { accepts: accepts, limit: limit, debug: debug });

          case 25:
            files = _context.sent;

            if (debug) console.timeEnd('> [debug] locating files ' + path);

            // always include manifest as npm does not allow ignoring it
            // source: https://docs.npmjs.com/files/package.json#files
            files.push(asAbsolute('package.json', path));

            // get files
            return _context.abrupt('return', (0, _arrayUnique2.default)(files));

          case 29:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function npm(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
}();

/**
 * Returns a list of files in the given
 * directory that are subject to be
 * sent to docker as build context.
 *
 * @param {String} full path to directory
 * @param {String} contents of `Dockerfile`
 * @param {Object} options:
 *  - `limit` {Number|null} byte limit
 *  - `debug` {Boolean} warn upon ignore
 * @return {Array} comprehensive list of paths to sync
 */

var docker = exports.docker = function () {
  var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(path) {
    var _ref4 = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var _ref4$limit = _ref4.limit;
    var limit = _ref4$limit === undefined ? null : _ref4$limit;
    var _ref4$debug = _ref4.debug;
    var debug = _ref4$debug === undefined ? false : _ref4$debug;
    var search_, search, filter, prefixLength, accepts, files;
    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            // base search path
            search_ = ['.'];

            // convert all filenames into absolute paths

            search = search_.map(function (file) {
              return asAbsolute(file, path);
            });

            // compile list of ignored patterns and files

            _context2.t0 = (0, _ignore2.default)();
            _context2.t1 = _ignored2.default + '\n';
            _context2.next = 6;
            return maybeRead((0, _path.resolve)(path, '.dockerignore'));

          case 6:
            _context2.t2 = _context2.sent;
            _context2.t3 = _context2.t1 + _context2.t2;
            filter = _context2.t0.add.call(_context2.t0, _context2.t3).createFilter();
            prefixLength = path.length + 1;

            accepts = function accepts(file) {
              var relativePath = file.substr(prefixLength);
              if ('' === relativePath) return true;
              var accepted = filter(relativePath);
              if (!accepted && debug) {
                console.log('> [debug] ignoring "%s"', file);
              }
              return accepted;
            };

            // locate files


            if (debug) console.time('> [debug] locating files ' + path);
            _context2.next = 14;
            return explode(search, { accepts: accepts, limit: limit, debug: debug });

          case 14:
            files = _context2.sent;

            if (debug) console.timeEnd('> [debug] locating files ' + path);

            // always include manifest as npm does not allow ignoring it
            // source: https://docs.npmjs.com/files/package.json#files
            files.push(asAbsolute('Dockerfile', path));

            // get files
            return _context2.abrupt('return', (0, _arrayUnique2.default)(files));

          case 18:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function docker(_x5, _x6) {
    return _ref3.apply(this, arguments);
  };
}();

/**
 * Transform relative paths into absolutes,
 * and maintains absolutes as such.
 *
 * @param {String} maybe relative path
 * @param {String} parent full path
 */

var _arrFlatten = require('arr-flatten');

var _arrFlatten2 = _interopRequireDefault(_arrFlatten);

var _arrayUnique = require('array-unique');

var _arrayUnique2 = _interopRequireDefault(_arrayUnique);

var _ignored = require('./ignored');

var _ignored2 = _interopRequireDefault(_ignored);

var _path = require('path');

var _fsPromise = require('fs-promise');

var _ignore = require('ignore');

var _ignore2 = _interopRequireDefault(_ignore);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var asAbsolute = function asAbsolute(path, parent) {
  if ('/' === path[0]) return path;
  return (0, _path.resolve)(parent, path);
};

/**
 * Explodes directories into a full list of files.
 * Eg:
 *   in:  ['/a.js', '/b']
 *   out: ['/a.js', '/b/c.js', '/b/d.js']
 *
 * @param {Array} of {String}s representing paths
 * @param {Array} of ignored {String}s.
 * @param {Object} options:
 *  - `limit` {Number|null} byte limit
 *  - `debug` {Boolean} warn upon ignore
 * @return {Array} of {String}s of full paths
 */

var explode = function () {
  var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee6(paths, _ref6) {
    var _this = this;

    var accepts = _ref6.accepts;
    var limit = _ref6.limit;
    var debug = _ref6.debug;
    var many, list;
    return _regenerator2.default.wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            many = function () {
              var _ref7 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(all) {
                return _regenerator2.default.wrap(function _callee4$(_context4) {
                  while (1) {
                    switch (_context4.prev = _context4.next) {
                      case 0:
                        _context4.next = 2;
                        return _promise2.default.all(all.map(function () {
                          var _ref8 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(file) {
                            return _regenerator2.default.wrap(function _callee3$(_context3) {
                              while (1) {
                                switch (_context3.prev = _context3.next) {
                                  case 0:
                                    _context3.next = 2;
                                    return list(file);

                                  case 2:
                                    return _context3.abrupt('return', _context3.sent);

                                  case 3:
                                  case 'end':
                                    return _context3.stop();
                                }
                              }
                            }, _callee3, _this);
                          }));

                          return function (_x11) {
                            return _ref8.apply(this, arguments);
                          };
                        }()));

                      case 2:
                        return _context4.abrupt('return', _context4.sent);

                      case 3:
                      case 'end':
                        return _context4.stop();
                    }
                  }
                }, _callee4, _this);
              }));

              return function many(_x10) {
                return _ref7.apply(this, arguments);
              };
            }();

            list = function () {
              var _ref9 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(file) {
                var path, s, all;
                return _regenerator2.default.wrap(function _callee5$(_context5) {
                  while (1) {
                    switch (_context5.prev = _context5.next) {
                      case 0:
                        path = file;
                        s = void 0;

                        if (accepts(file)) {
                          _context5.next = 4;
                          break;
                        }

                        return _context5.abrupt('return', null);

                      case 4:
                        _context5.prev = 4;
                        _context5.next = 7;
                        return (0, _fsPromise.stat)(path);

                      case 7:
                        s = _context5.sent;
                        _context5.next = 22;
                        break;

                      case 10:
                        _context5.prev = 10;
                        _context5.t0 = _context5['catch'](4);

                        // in case the file comes from `files` or `main`
                        // and it wasn't specified with `.js` by the user
                        path = file + '.js';

                        _context5.prev = 13;
                        _context5.next = 16;
                        return (0, _fsPromise.stat)(path);

                      case 16:
                        s = _context5.sent;
                        _context5.next = 22;
                        break;

                      case 19:
                        _context5.prev = 19;
                        _context5.t1 = _context5['catch'](13);
                        return _context5.abrupt('return', null);

                      case 22:
                        if (!s.isDirectory()) {
                          _context5.next = 29;
                          break;
                        }

                        _context5.next = 25;
                        return (0, _fsPromise.readdir)(file);

                      case 25:
                        all = _context5.sent;
                        return _context5.abrupt('return', many(all.map(function (subdir) {
                          return asAbsolute(subdir, file);
                        })));

                      case 29:
                        return _context5.abrupt('return', path);

                      case 30:
                      case 'end':
                        return _context5.stop();
                    }
                  }
                }, _callee5, _this, [[4, 10], [13, 19]]);
              }));

              return function list(_x12) {
                return _ref9.apply(this, arguments);
              };
            }();

            _context6.next = 4;
            return many(paths);

          case 4:
            _context6.t0 = _context6.sent;

            _context6.t1 = function (v) {
              return null !== v;
            };

            return _context6.abrupt('return', (0, _arrFlatten2.default)(_context6.t0).filter(_context6.t1));

          case 7:
          case 'end':
            return _context6.stop();
        }
      }
    }, _callee6, this);
  }));

  return function explode(_x8, _x9) {
    return _ref5.apply(this, arguments);
  };
}();

/**
 * Returns the contents of a file if it exists.
 *
 * @return {String} results or `''`
 */

var maybeRead = function () {
  var _ref10 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee7(path) {
    var default_ = arguments.length <= 1 || arguments[1] === undefined ? '' : arguments[1];
    return _regenerator2.default.wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            _context7.prev = 0;
            _context7.next = 3;
            return (0, _fsPromise.readFile)(path, 'utf8');

          case 3:
            return _context7.abrupt('return', _context7.sent);

          case 6:
            _context7.prev = 6;
            _context7.t0 = _context7['catch'](0);
            return _context7.abrupt('return', default_);

          case 9:
          case 'end':
            return _context7.stop();
        }
      }
    }, _callee7, this, [[0, 6]]);
  }));

  return function maybeRead(_x13, _x14) {
    return _ref10.apply(this, arguments);
  };
}();

/**
 * Remove leading `./` from the beginning of ignores
 * because our parser doesn't like them :|
 */

var clearRelative = function clearRelative(str) {
  return str.replace(/(\n|^)\.\//g, '$1');
};
