// now/build/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _from = require('babel-runtime/core-js/array/from');

var _from2 = _interopRequireDefault(_from);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _bytes = require('bytes');

var _bytes2 = _interopRequireDefault(_bytes);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _getFiles = require('./get-files');

var _ua = require('./ua');

var _ua2 = _interopRequireDefault(_ua);

var _hash = require('./hash');

var _hash2 = _interopRequireDefault(_hash);

var _asyncRetry = require('async-retry');

var _asyncRetry2 = _interopRequireDefault(_asyncRetry);

var _agent = require('./agent');

var _agent2 = _interopRequireDefault(_agent);

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _path = require('path');

var _os = require('os');

var _ini = require('ini');

var _fsPromise = require('fs-promise');

var _resumer = require('resumer');

var _resumer2 = _interopRequireDefault(_resumer);

var _splitArray = require('split-array');

var _splitArray2 = _interopRequireDefault(_splitArray);

var _dockerFileParser = require('docker-file-parser');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// how many concurrent HTTP/2 stream uploads
var MAX_CONCURRENT = 10;

// check if running windows
var IS_WIN = /^win/.test(process.platform);
var SEP = IS_WIN ? '\\' : '/';

var Now = function (_EventEmitter) {
  (0, _inherits3.default)(Now, _EventEmitter);

  function Now(url, token, _ref) {
    var _ref$forceNew = _ref.forceNew;
    var forceNew = _ref$forceNew === undefined ? false : _ref$forceNew;
    var _ref$debug = _ref.debug;
    var debug = _ref$debug === undefined ? false : _ref$debug;
    (0, _classCallCheck3.default)(this, Now);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Now.__proto__ || (0, _getPrototypeOf2.default)(Now)).call(this));

    _this._token = token;
    _this._debug = debug;
    _this._forceNew = forceNew;
    _this._agent = new _agent2.default(url, { debug: debug });
    _this._onRetry = _this._onRetry.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(Now, [{
    key: 'create',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(path, _ref3) {
        var _this2 = this;

        var wantsPublic = _ref3.wantsPublic;
        var _ref3$quiet = _ref3.quiet;
        var quiet = _ref3$quiet === undefined ? false : _ref3$quiet;
        var _ref3$env = _ref3.env;
        var env = _ref3$env === undefined ? {} : _ref3$env;
        var _ref3$forceNew = _ref3.forceNew;
        var forceNew = _ref3$forceNew === undefined ? false : _ref3$forceNew;
        var _ref3$forceSync = _ref3.forceSync;
        var forceSync = _ref3$forceSync === undefined ? false : _ref3$forceSync;
        var _ref3$forwardNpm = _ref3.forwardNpm;
        var forwardNpm = _ref3$forwardNpm === undefined ? false : _ref3$forwardNpm;
        var _ref3$deploymentType = _ref3.deploymentType;
        var deploymentType = _ref3$deploymentType === undefined ? 'npm' : _ref3$deploymentType;

        var pkg, name, description, files, e, _e, nowProperties, npmrc, authToken, hashes, engines, deployment, missingVersion;

        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                this._path = path;

                pkg = {};
                name = void 0, description = void 0, files = void 0;

                if (!('npm' === deploymentType)) {
                  _context3.next = 29;
                  break;
                }

                _context3.prev = 4;
                _context3.next = 7;
                return (0, _fsPromise.readFile)((0, _path.resolve)(path, 'package.json'));

              case 7:
                pkg = _context3.sent;

                pkg = JSON.parse(pkg);
                _context3.next = 16;
                break;

              case 11:
                _context3.prev = 11;
                _context3.t0 = _context3['catch'](4);
                e = Error('Failed to read JSON in "' + path + '/package.json"');

                e.userError = true;
                throw e;

              case 16:
                if (!(!pkg.scripts || !pkg.scripts.start && !pkg.scripts['now-start'])) {
                  _context3.next = 20;
                  break;
                }

                _e = Error('Missing `start` (or `now-start`) script in `package.json`. ' + 'See: https://docs.npmjs.com/cli/start.');

                _e.userError = true;
                throw _e;

              case 20:

                if (null == pkg.name || 'string' !== typeof pkg.name) {
                  name = (0, _path.basename)(path);
                  if (!quiet) console.log('> No `name` in `package.json`, using ' + _chalk2.default.bold(name));
                } else {
                  name = pkg.name;
                }

                description = pkg.description;

                if (this._debug) console.time('> [debug] Getting files');
                _context3.next = 25;
                return (0, _getFiles.npm)(path, pkg, { debug: this._debug });

              case 25:
                files = _context3.sent;

                if (this._debug) console.timeEnd('> [debug] Getting files');
                _context3.next = 31;
                break;

              case 29:
                if (!('docker' === deploymentType)) {
                  _context3.next = 31;
                  break;
                }

                return _context3.delegateYield(_regenerator2.default.mark(function _callee() {
                  var docker, dockerfile, _e2, _e3, _e4, _e5, labels;

                  return _regenerator2.default.wrap(function _callee$(_context) {
                    while (1) {
                      switch (_context.prev = _context.next) {
                        case 0:
                          docker = void 0;
                          _context.prev = 1;
                          _context.next = 4;
                          return (0, _fsPromise.readFile)((0, _path.resolve)(path, 'Dockerfile'), 'utf8');

                        case 4:
                          dockerfile = _context.sent;

                          docker = (0, _dockerFileParser.parse)(dockerfile, { includeComments: true });
                          _context.next = 13;
                          break;

                        case 8:
                          _context.prev = 8;
                          _context.t0 = _context['catch'](1);
                          _e2 = Error('Failed to parse "' + path + '/Dockerfile"');

                          _e2.userError = true;
                          throw _e2;

                        case 13:
                          if (docker.length) {
                            _context.next = 17;
                            break;
                          }

                          _e3 = Error('No commands found in `Dockerfile`');

                          _e3.userError = true;
                          throw _e3;

                        case 17:
                          if (docker.some(function (cmd) {
                            return 'CMD' === cmd.name;
                          })) {
                            _context.next = 21;
                            break;
                          }

                          _e4 = Error('No `CMD` found in `Dockerfile`. ' + 'See: https://docs.docker.com/engine/reference/builder/#/cmd');

                          _e4.userError = true;
                          throw _e4;

                        case 21:
                          if (docker.some(function (cmd) {
                            return 'EXPOSE' === cmd.name;
                          })) {
                            _context.next = 25;
                            break;
                          }

                          _e5 = Error('No `EXPOSE` found in `Dockerfile`. A port must be supplied. ' + 'See: https://docs.docker.com/engine/reference/builder/#/expose');

                          _e5.userError = true;
                          throw _e5;

                        case 25:
                          labels = {};

                          docker.filter(function (cmd) {
                            return 'LABEL' === cmd.name;
                          }).forEach(function (_ref4) {
                            var args = _ref4.args;

                            for (var key in args) {
                              // unescape and convert into string
                              try {
                                labels[key] = JSON.parse(args[key]);
                              } catch (err) {
                                var _e6 = Error('Error parsing value for LABEL ' + key + ' in `Dockerfile`');
                                _e6.userError = true;
                                throw _e6;
                              }
                            }
                          });

                          if (null == labels.name) {
                            name = (0, _path.basename)(path);
                            if (!quiet) console.log('> No `name` LABEL in `Dockerfile`, using ' + _chalk2.default.bold(name));
                          } else {
                            name = labels.name;
                          }

                          description = labels.description;

                          if (_this2._debug) console.time('> [debug] Getting files');
                          _context.next = 32;
                          return (0, _getFiles.docker)(path, { debug: _this2._debug });

                        case 32:
                          files = _context.sent;

                          if (_this2._debug) console.timeEnd('> [debug] Getting files');

                        case 34:
                        case 'end':
                          return _context.stop();
                      }
                    }
                  }, _callee, _this2, [[1, 8]]);
                })(), 't1', 31);

              case 31:
                nowProperties = pkg ? pkg.now || {} : {};


                forwardNpm = forwardNpm || nowProperties['forwardNpm'];

                // Read .npmrc
                npmrc = {};
                authToken = void 0;

                if (!('npm' === deploymentType && forwardNpm)) {
                  _context3.next = 57;
                  break;
                }

                _context3.prev = 36;
                _context3.next = 39;
                return (0, _fsPromise.readFile)((0, _path.resolve)(path, '.npmrc'), 'utf8');

              case 39:
                npmrc = _context3.sent;

                npmrc = (0, _ini.parse)(npmrc);
                authToken = npmrc['//registry.npmjs.org/:_authToken'];
                _context3.next = 46;
                break;

              case 44:
                _context3.prev = 44;
                _context3.t2 = _context3['catch'](36);

              case 46:
                if (authToken) {
                  _context3.next = 57;
                  break;
                }

                _context3.prev = 47;
                _context3.next = 50;
                return (0, _fsPromise.readFile)((0, _path.resolve)((0, _os.homedir)(), '.npmrc'), 'utf8');

              case 50:
                npmrc = _context3.sent;

                npmrc = (0, _ini.parse)(npmrc);
                authToken = npmrc['//registry.npmjs.org/:_authToken'];
                _context3.next = 57;
                break;

              case 55:
                _context3.prev = 55;
                _context3.t3 = _context3['catch'](47);

              case 57:

                if (this._debug) console.time('> [debug] Computing hashes');
                _context3.next = 60;
                return (0, _hash2.default)(files);

              case 60:
                hashes = _context3.sent;

                if (this._debug) console.timeEnd('> [debug] Computing hashes');

                this._files = hashes;

                engines = nowProperties.engines || pkg.engines;
                _context3.next = 66;
                return this.retry(function () {
                  var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(bail) {
                    var res, body, err;
                    return _regenerator2.default.wrap(function _callee2$(_context2) {
                      while (1) {
                        switch (_context2.prev = _context2.next) {
                          case 0:
                            if (_this2._debug) console.time('> [debug] /now/create');
                            _context2.next = 3;
                            return _this2._fetch('/now/create', {
                              method: 'POST',
                              body: {
                                env: env,
                                public: wantsPublic,
                                forceNew: forceNew,
                                forceSync: forceSync,
                                name: name,
                                description: description,
                                deploymentType: deploymentType,
                                registryAuthToken: authToken,
                                // Flatten the array to contain files to sync where each nested input
                                // array has a group of files with the same sha but different path
                                files: Array.prototype.concat.apply([], (0, _from2.default)(_this2._files).map(function (_ref6) {
                                  var _ref7 = (0, _slicedToArray3.default)(_ref6, 2);

                                  var sha = _ref7[0];
                                  var _ref7$ = _ref7[1];
                                  var data = _ref7$.data;
                                  var names = _ref7$.names;

                                  return names.map(function (n) {
                                    return {
                                      sha: sha,
                                      size: data.length,
                                      file: toRelative(n, _this2._path)
                                    };
                                  });
                                })),
                                engines: engines
                              }
                            });

                          case 3:
                            res = _context2.sent;

                            if (_this2._debug) console.timeEnd('> [debug] /now/create');

                            // no retry on 4xx
                            body = void 0;
                            _context2.prev = 6;
                            _context2.next = 9;
                            return res.json();

                          case 9:
                            body = _context2.sent;
                            _context2.next = 15;
                            break;

                          case 12:
                            _context2.prev = 12;
                            _context2.t0 = _context2['catch'](6);
                            throw new Error('Unexpected response');

                          case 15:
                            if (!(429 === res.status)) {
                              _context2.next = 19;
                              break;
                            }

                            return _context2.abrupt('return', bail(responseError(res)));

                          case 19:
                            if (!(res.status >= 400 && res.status < 500)) {
                              _context2.next = 25;
                              break;
                            }

                            err = new Error(body.error.message);

                            err.userError = true;
                            return _context2.abrupt('return', bail(err));

                          case 25:
                            if (!(200 !== res.status)) {
                              _context2.next = 27;
                              break;
                            }

                            throw new Error(body.error.message);

                          case 27:
                            return _context2.abrupt('return', body);

                          case 28:
                          case 'end':
                            return _context2.stop();
                        }
                      }
                    }, _callee2, _this2, [[6, 12]]);
                  }));

                  return function (_x3) {
                    return _ref5.apply(this, arguments);
                  };
                }());

              case 66:
                deployment = _context3.sent;


                // we report about files whose sizes are too big
                missingVersion = false;

                if (deployment.warnings) {
                  (function () {
                    var sizeExceeded = 0;
                    deployment.warnings.forEach(function (warning) {
                      if ('size_limit_exceeded' === warning.reason) {
                        var sha = warning.sha;
                        var limit = warning.limit;

                        var n = hashes.get(sha).names.pop();
                        console.error('> \u001b[31mWarning!\u001b[39m Skipping file %s (size exceeded %s)', n, (0, _bytes2.default)(limit));
                        hashes.get(sha).names.unshift(n); // move name (hack, if duplicate matches we report them in order)
                        sizeExceeded++;
                      } else if ('node_version_not_found' === warning.reason) {
                        var wanted = warning.wanted;
                        var used = warning.used;

                        console.error('> \u001b[31mWarning!\u001b[39m Requested node version %s is not available', wanted, used);
                        missingVersion = true;
                      }
                    });

                    if (sizeExceeded) {
                      console.error('> \u001b[31mWarning!\u001b[39m ' + sizeExceeded + ' of the files ' + 'exceeded the limit for your plan.\n' + ('> See ' + _chalk2.default.underline('https://zeit.co/account') + ' to upgrade.'));
                    }
                  })();
                }

                if (!quiet && deployment.nodeVersion) {
                  if (engines && engines.node) {
                    if (missingVersion) {
                      console.log('> Using Node.js ' + _chalk2.default.bold(deployment.nodeVersion) + ' (default)');
                    } else {
                      console.log('> Using Node.js ' + _chalk2.default.bold(deployment.nodeVersion) + ' (requested: ' + _chalk2.default.dim('`' + engines.node + '`') + ')');
                    }
                  } else {
                    console.log('> Using Node.js ' + _chalk2.default.bold(deployment.nodeVersion) + ' (default)');
                  }
                }

                this._id = deployment.deploymentId;
                this._host = deployment.url;
                this._missing = deployment.missing || [];

                return _context3.abrupt('return', this._url);

              case 74:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this, [[4, 11], [36, 44], [47, 55]]);
      }));

      function create(_x, _x2) {
        return _ref2.apply(this, arguments);
      }

      return create;
    }()
  }, {
    key: 'upload',
    value: function upload() {
      var _this3 = this;

      var parts = (0, _splitArray2.default)(this._missing, MAX_CONCURRENT);

      if (this._debug) {
        console.log('> [debug] Will upload ' + (this._missing.length + ' files in ' + parts.length + ' ') + ('steps of ' + MAX_CONCURRENT + ' uploads.'));
      }

      var uploadChunk = function uploadChunk() {
        _promise2.default.all(parts.shift().map(function (sha) {
          return (0, _asyncRetry2.default)(function () {
            var _ref8 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(bail, attempt) {
              var file, data, names, stream, res;
              return _regenerator2.default.wrap(function _callee4$(_context4) {
                while (1) {
                  switch (_context4.prev = _context4.next) {
                    case 0:
                      file = _this3._files.get(sha);
                      data = file.data;
                      names = file.names;


                      if (_this3._debug) console.time('> [debug] /sync #' + attempt + ' ' + names.join(' '));
                      stream = (0, _resumer2.default)().queue(data).end();
                      _context4.next = 7;
                      return _this3._fetch('/now/sync', {
                        method: 'POST',
                        headers: {
                          'Content-Type': 'application/octet-stream',
                          'Content-Length': data.length,
                          'x-now-deployment-id': _this3._id,
                          'x-now-sha': sha,
                          'x-now-file': names.map(function (name) {
                            return toRelative(name, _this3._path);
                          }).join(','),
                          'x-now-size': data.length
                        },
                        body: stream
                      });

                    case 7:
                      res = _context4.sent;

                      if (_this3._debug) console.timeEnd('> [debug] /sync #' + attempt + ' ' + names.join(' '));

                      // no retry on 4xx

                      if (!(200 !== res.status && (400 <= res.status || 500 > res.status))) {
                        _context4.next = 12;
                        break;
                      }

                      if (_this3._debug) console.log('> [debug] bailing on creating due to %s', res.status);
                      return _context4.abrupt('return', bail(responseError(res)));

                    case 12:

                      _this3.emit('upload', file);

                    case 13:
                    case 'end':
                      return _context4.stop();
                  }
                }
              }, _callee4, _this3);
            }));

            return function (_x4, _x5) {
              return _ref8.apply(this, arguments);
            };
          }(), { retries: 3, randomize: true, onRetry: _this3._onRetry });
        })).then(function () {
          return parts.length ? uploadChunk() : _this3.emit('complete');
        }).catch(function (err) {
          return _this3.emit('error', err);
        });
      };

      uploadChunk();
    }
  }, {
    key: 'listSecrets',
    value: function () {
      var _ref9 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee6() {
        var _this4 = this;

        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                return _context6.abrupt('return', this.retry(function () {
                  var _ref10 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(bail, attempt) {
                    var res, body;
                    return _regenerator2.default.wrap(function _callee5$(_context5) {
                      while (1) {
                        switch (_context5.prev = _context5.next) {
                          case 0:
                            if (_this4._debug) console.time('> [debug] #' + attempt + ' GET /secrets');
                            _context5.next = 3;
                            return _this4._fetch('/now/secrets');

                          case 3:
                            res = _context5.sent;

                            if (_this4._debug) console.timeEnd('> [debug] #' + attempt + ' GET /secrets');
                            _context5.next = 7;
                            return res.json();

                          case 7:
                            body = _context5.sent;
                            return _context5.abrupt('return', body.secrets);

                          case 9:
                          case 'end':
                            return _context5.stop();
                        }
                      }
                    }, _callee5, _this4);
                  }));

                  return function (_x6, _x7) {
                    return _ref10.apply(this, arguments);
                  };
                }()));

              case 1:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function listSecrets() {
        return _ref9.apply(this, arguments);
      }

      return listSecrets;
    }()
  }, {
    key: 'list',
    value: function () {
      var _ref11 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee8(app) {
        var _this5 = this;

        var query, _ref12, deployments;

        return _regenerator2.default.wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                query = app ? '?app=' + encodeURIComponent(app) : '';
                _context8.next = 3;
                return this.retry(function () {
                  var _ref13 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee7(bail) {
                    var res;
                    return _regenerator2.default.wrap(function _callee7$(_context7) {
                      while (1) {
                        switch (_context7.prev = _context7.next) {
                          case 0:
                            if (_this5._debug) console.time('> [debug] /list');
                            _context7.next = 3;
                            return _this5._fetch('/now/list' + query);

                          case 3:
                            res = _context7.sent;

                            if (_this5._debug) console.timeEnd('> [debug] /list');

                            // no retry on 4xx

                            if (!(400 <= res.status && 500 > res.status)) {
                              _context7.next = 8;
                              break;
                            }

                            if (_this5._debug) {
                              console.log('> [debug] bailing on listing due to %s', res.status);
                            }
                            return _context7.abrupt('return', bail(responseError(res)));

                          case 8:
                            if (!(200 !== res.status)) {
                              _context7.next = 10;
                              break;
                            }

                            throw new Error('Fetching deployment list failed');

                          case 10:
                            return _context7.abrupt('return', res.json());

                          case 11:
                          case 'end':
                            return _context7.stop();
                        }
                      }
                    }, _callee7, _this5);
                  }));

                  return function (_x9) {
                    return _ref13.apply(this, arguments);
                  };
                }(), { retries: 3, minTimeout: 2500, onRetry: this._onRetry });

              case 3:
                _ref12 = _context8.sent;
                deployments = _ref12.deployments;
                return _context8.abrupt('return', deployments);

              case 6:
              case 'end':
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function list(_x8) {
        return _ref11.apply(this, arguments);
      }

      return list;
    }()
  }, {
    key: 'listAliases',
    value: function () {
      var _ref14 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee10(deploymentId) {
        var _this6 = this;

        return _regenerator2.default.wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                return _context10.abrupt('return', this.retry(function () {
                  var _ref15 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee9(bail, attempt) {
                    var res, body;
                    return _regenerator2.default.wrap(function _callee9$(_context9) {
                      while (1) {
                        switch (_context9.prev = _context9.next) {
                          case 0:
                            _context9.next = 2;
                            return _this6._fetch(deploymentId ? '/now/deployments/' + deploymentId + '/aliases' : '/now/aliases');

                          case 2:
                            res = _context9.sent;
                            _context9.next = 5;
                            return res.json();

                          case 5:
                            body = _context9.sent;
                            return _context9.abrupt('return', body.aliases);

                          case 7:
                          case 'end':
                            return _context9.stop();
                        }
                      }
                    }, _callee9, _this6);
                  }));

                  return function (_x11, _x12) {
                    return _ref15.apply(this, arguments);
                  };
                }()));

              case 1:
              case 'end':
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function listAliases(_x10) {
        return _ref14.apply(this, arguments);
      }

      return listAliases;
    }()
  }, {
    key: 'getNameservers',
    value: function getNameservers(domain) {
      var _this7 = this;

      return new _promise2.default(function (resolve, reject) {
        var fallback = false;

        _this7.retry(function () {
          var _ref16 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee11(bail, attempt) {
            var res, body;
            return _regenerator2.default.wrap(function _callee11$(_context11) {
              while (1) {
                switch (_context11.prev = _context11.next) {
                  case 0:
                    if (_this7._debug) console.time('> [debug] #' + attempt + ' GET /whois-ns' + (fallback ? '-fallback' : ''));
                    _context11.next = 3;
                    return _this7._fetch('/whois-ns' + (fallback ? '-fallback' : '') + '?domain=' + encodeURIComponent(domain));

                  case 3:
                    res = _context11.sent;

                    if (_this7._debug) console.timeEnd('> [debug] #' + attempt + ' GET /whois-ns' + (fallback ? '-fallback' : ''));
                    _context11.next = 7;
                    return res.json();

                  case 7:
                    body = _context11.sent;

                    if (!(200 === res.status)) {
                      _context11.next = 15;
                      break;
                    }

                    if (!((!body.nameservers || body.nameservers.length === 0) && !fallback)) {
                      _context11.next = 12;
                      break;
                    }

                    // if the nameservers are `null` it's likely
                    // that our whois service failed to parse it
                    fallback = true;
                    throw new Error('Invalid whois response');

                  case 12:
                    return _context11.abrupt('return', body);

                  case 15:
                    if (attempt > 1) fallback = true;
                    throw new Error('Whois error (' + res.status + '): ' + body.error.message);

                  case 17:
                  case 'end':
                    return _context11.stop();
                }
              }
            }, _callee11, _this7);
          }));

          return function (_x13, _x14) {
            return _ref16.apply(this, arguments);
          };
        }()).then(function (body) {
          body.nameservers = body.nameservers.filter(function (ns) {
            // temporary hack:
            // sometimes we get a response that looks like:
            // ['ns', 'ns', '', '']
            // so we filter the empty ones
            return ns.length;
          });
          resolve(body);
        });
      });
    }

    // _ensures_ the domain is setup (idempotent)

  }, {
    key: 'setupDomain',
    value: function setupDomain(name) {
      var _this8 = this;

      var _ref17 = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

      var isExternal = _ref17.isExternal;

      return this.retry(function () {
        var _ref18 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee12(bail, attempt) {
          var res, _body, code, err, body;

          return _regenerator2.default.wrap(function _callee12$(_context12) {
            while (1) {
              switch (_context12.prev = _context12.next) {
                case 0:
                  if (_this8._debug) console.time('> [debug] #' + attempt + ' POST /domains');
                  _context12.next = 3;
                  return _this8._fetch('/domains', {
                    method: 'POST',
                    body: { name: name, isExternal: !!isExternal }
                  });

                case 3:
                  res = _context12.sent;

                  if (_this8._debug) console.timeEnd('> [debug] #' + attempt + ' POST /domains');

                  if (!(403 === res.status)) {
                    _context12.next = 14;
                    break;
                  }

                  _context12.next = 8;
                  return res.json();

                case 8:
                  _body = _context12.sent;
                  code = _body.error.code;
                  err = void 0;


                  if ('custom_domain_needs_upgrade' === code) {
                    err = new Error('Custom domains are only enabled for premium accounts. Please upgrade at ' + _chalk2.default.underline('https://zeit.co/account') + '.');
                  } else {
                    err = new Error('Not authorized to access domain ' + name);
                  }

                  err.userError = true;
                  return _context12.abrupt('return', bail(err));

                case 14:
                  _context12.next = 16;
                  return res.json();

                case 16:
                  body = _context12.sent;

                  if (!(409 === res.status)) {
                    _context12.next = 20;
                    break;
                  }

                  if (_this8._debug) console.log('> [debug] Domain already exists (noop)');
                  return _context12.abrupt('return', { uid: body.error.uid });

                case 20:
                  if (!(200 !== res.status)) {
                    _context12.next = 22;
                    break;
                  }

                  throw new Error(body.error.message);

                case 22:
                  return _context12.abrupt('return', body);

                case 23:
                case 'end':
                  return _context12.stop();
              }
            }
          }, _callee12, _this8);
        }));

        return function (_x16, _x17) {
          return _ref18.apply(this, arguments);
        };
      }());
    }
  }, {
    key: 'remove',
    value: function () {
      var _ref19 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee14(deploymentId, _ref20) {
        var _this9 = this;

        var hard = _ref20.hard;
        var data;
        return _regenerator2.default.wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                data = { deploymentId: deploymentId, hard: hard };
                _context14.next = 3;
                return this.retry(function () {
                  var _ref21 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee13(bail) {
                    var res;
                    return _regenerator2.default.wrap(function _callee13$(_context13) {
                      while (1) {
                        switch (_context13.prev = _context13.next) {
                          case 0:
                            if (_this9._debug) console.time('> [debug] /remove');
                            _context13.next = 3;
                            return _this9._fetch('/now/remove', {
                              method: 'DELETE',
                              body: data
                            });

                          case 3:
                            res = _context13.sent;

                            if (_this9._debug) console.timeEnd('> [debug] /remove');

                            // no retry on 4xx

                            if (!(400 <= res.status && 500 > res.status)) {
                              _context13.next = 8;
                              break;
                            }

                            if (_this9._debug) {
                              console.log('> [debug] bailing on removal due to %s', res.status);
                            }
                            return _context13.abrupt('return', bail(responseError(res)));

                          case 8:
                            if (!(200 !== res.status)) {
                              _context13.next = 10;
                              break;
                            }

                            throw new Error('Removing deployment failed');

                          case 10:
                          case 'end':
                            return _context13.stop();
                        }
                      }
                    }, _callee13, _this9);
                  }));

                  return function (_x20) {
                    return _ref21.apply(this, arguments);
                  };
                }());

              case 3:
                return _context14.abrupt('return', true);

              case 4:
              case 'end':
                return _context14.stop();
            }
          }
        }, _callee14, this);
      }));

      function remove(_x18, _x19) {
        return _ref19.apply(this, arguments);
      }

      return remove;
    }()
  }, {
    key: 'retry',
    value: function retry(fn) {
      var _ref22 = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

      var _ref22$retries = _ref22.retries;
      var retries = _ref22$retries === undefined ? 3 : _ref22$retries;
      var _ref22$maxTimeout = _ref22.maxTimeout;
      var maxTimeout = _ref22$maxTimeout === undefined ? Infinity : _ref22$maxTimeout;

      return (0, _asyncRetry2.default)(fn, {
        retries: retries,
        maxTimeout: maxTimeout,
        onRetry: this._onRetry
      });
    }
  }, {
    key: '_onRetry',
    value: function _onRetry(err) {
      if (this._debug) {
        console.log('> [debug] Retrying: ' + err.stack);
      }
    }
  }, {
    key: 'close',
    value: function close() {
      this._agent.close();
    }
  }, {
    key: '_fetch',
    value: function _fetch(_url) {
      var opts = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

      opts.headers = opts.headers || {};
      opts.headers.authorization = 'Bearer ' + this._token;
      opts.headers['user-agent'] = _ua2.default;
      return this._agent.fetch(_url, opts);
    }
  }, {
    key: 'id',
    get: function get() {
      return this._id;
    }
  }, {
    key: 'url',
    get: function get() {
      return 'https://' + this._host;
    }
  }, {
    key: 'host',
    get: function get() {
      return this._host;
    }
  }, {
    key: 'syncAmount',
    get: function get() {
      var _this10 = this;

      if (!this._syncAmount) {
        this._syncAmount = this._missing.map(function (sha) {
          return _this10._files.get(sha).data.length;
        }).reduce(function (a, b) {
          return a + b;
        }, 0);
      }
      return this._syncAmount;
    }
  }]);
  return Now;
}(_events2.default);

exports.default = Now;


function toRelative(path, base) {
  var fullBase = base.endsWith(SEP) ? base : base + SEP;
  var relative = path.substr(fullBase.length);
  if (relative.startsWith(SEP)) relative = relative.substr(1);
  return relative.replace(/\\/g, '/');
}

function responseError(res) {
  var err = new Error('Response error');
  err.status = res.status;

  if (429 === res.status) {
    var retryAfter = res.headers.get('Retry-After');
    if (retryAfter) {
      err.retryAfter = parseInt(retryAfter, 10);
    }
  }

  return err;
}
